StyleCop.Analyzers/StyleCop.Analyzers/OrderingRules/SA1202ElementsMustBeOrderedByAccess.cs

```diff
@@ -125,39 +145,79 @@ private static void HandleMemberList(SyntaxNodeAnalysisContext context, SyntaxLi
-                {
-                    currentAccessLevel = AccessLevel.Public;
-                }
-                else
-                {
-                    currentAccessLevel = AccessLevelHelper.GetAccessLevel(modifiers);
-                    currentAccessLevel = currentAccessLevel == AccessLevel.NotSpecified ? defaultAccessLevel : currentAccessLevel;
-                }
+                AccessLevel currentAccessLevel = MemberOrderHelper.GetAccessLevelForOrdering(member, modifiers);
 
-                if (previousAccessLevel != AccessLevel.NotSpecified
-                    && currentSyntaxKind == previousSyntaxKind
-                    && currentAccessLevel > previousAccessLevel)
+                if (previousMember != null && previousAccessLevel != AccessLevel.NotSpecified)
                 {
-                    context.ReportDiagnostic(
-                        Diagnostic.Create(
-                            Descriptor,
-                            NamedTypeHelpers.GetNameOrIdentifierLocation(member),
-                            AccessLevelHelper.GetName(currentAccessLevel),
-                            MemberNames[currentSyntaxKind],
-                            AccessLevelHelper.GetName(previousAccessLevel)));
+                    bool compareAccessLevel = true;
+                    for (int j = 0; compareAccessLevel && j < accessibilityIndex; j++)
+                    {
+                        switch (elementOrder[j])
+                        {
+                        case OrderingTrait.Kind:
+                            if (previousSyntaxKind != currentSyntaxKind)
+                            {
+                                compareAccessLevel = false;
+                            }
+
+                            continue;
+
+                        case OrderingTrait.Constant:
+                            // Only fields may be marked const
+                            bool previousIsConst = previousMember.IsKind(SyntaxKind.FieldDeclaration) && previousMember.GetModifiers().Any(SyntaxKind.ConstKeyword);
+                            bool currentIsConst = member.IsKind(SyntaxKind.FieldDeclaration) && modifiers.Any(SyntaxKind.ConstKeyword);
+                            if (previousIsConst != currentIsConst)
+                            {
+                                compareAccessLevel = false;
+                            }
+
+                            continue;
+
+                        case OrderingTrait.Readonly:
+                            // Only fields may be marked readonly
+                            bool previousIsReadonly = previousMember.IsKind(SyntaxKind.FieldDeclaration) && previousMember.GetModifiers().Any(SyntaxKind.ReadOnlyKeyword);
+                            bool currentIsReadonly = member.IsKind(SyntaxKind.FieldDeclaration) && modifiers.Any(SyntaxKind.ReadOnlyKeyword);
```

Comment:
:exclamation: See remarks for `OrderingTrait.Constant` above. They apply here as well